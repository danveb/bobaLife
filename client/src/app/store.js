import { combineReducers, configureStore } from "@reduxjs/toolkit"; 
import authReducer from "../redux/auth/authSlice"; 
import productsReducer from "../redux/products/productsSlice"; 
import cartReducer from "../redux/cart/cartSlice"; 
import {
    persistStore,
    persistReducer,
    FLUSH,
    REHYDRATE,
    PAUSE,
    PERSIST,
    PURGE,
    REGISTER,
} from 'redux-persist';
import storage from "redux-persist/lib/storage"; 

// Perhaps combine all reducers and persist? 
// even persist user so we don't save user data in localStorage?
// combineReducers from rootReducer
const rootReducer = combineReducers({
    auth: authReducer, 
    products: productsReducer, 
    cart: cartReducer, 
}); 

const persistConfig = {
    key: 'root',
    storage,
}
  
// const persistedReducer = persistReducer(persistConfig, cartReducer)
const persistedReducer = persistReducer(persistConfig, rootReducer)

export const store = configureStore({
    // reducer: {
    //     auth: authReducer, 
    //     products: productsReducer, 
    //     cart: persistedReducer, 
    // },
    reducer: persistedReducer, 
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
            serializableCheck: {
                ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
            },
        }),
}); 

export const persistor = persistStore(store) 