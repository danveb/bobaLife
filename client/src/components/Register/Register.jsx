import { useState, useEffect } from "react"; 
import { Link, useNavigate } from "react-router-dom"; 
import { useSelector, useDispatch } from "react-redux"; 
import { toast } from "react-toastify"; 
import { register, reset } from "../../redux/auth/authSlice"; 
import { Navbar, Menu, Spinner } from "../index";
import "./Register.scss"; 

const Register = ({ menuOpen, setMenuOpen }) => {
    // useState 
    const [formData, setFormData] = useState({
        username: "", 
        email: "", 
        password: "",
    }); 

    // destructured form data
    const { username, email, password } = formData; 

    // useNavigate
    const navigate = useNavigate();  

    // useDispatch
    const dispatch = useDispatch(); 

    // select from state.auth
    const { user, isLoading, isError, isSuccess, message } = useSelector((state) => state.auth); 

    // useEffect 
    useEffect(() => {
        if(isError) {
            toast.error("Sorry, an error occured. Try again or create a new account.", {
                position: "top-center", 
                autoClose: 2000, 
                pauseOnHover: true, 
            }); 
        }
        if(isSuccess) {
            toast.success(`Welcome ${user.username}!`, {
                position: "top-center", 
                autoClose: 1500, 
                pauseOnHover: false, 
            }); 
            navigate("/")
        }
        // redirect existing user back home 
        if(user) {
            navigate("/"); 
        }
        // dispatch reset() 
        dispatch(reset()); 

        // dependency array to fire useEffect if there are any changes 
    }, [user, isError, isSuccess, message, navigate, dispatch]);

    // handleChange
    const handleChange = (e) => {
        setFormData((prevState) => ({
            ...prevState, 
            [e.target.name]: e.target.value
        })); 
    }; 

    // handleSubmit
    const handleSubmit = (e) => {
        e.preventDefault(); 
        if(!password) {
            toast.error("Please enter password")
        } else {
            const userData = {
                username, 
                email, 
                password,
            }
            // error checking with toast... 
            if(username.length === 0) {
                toast.error("Username is empty. Please create it again."); 
            } else if(email.length === 0) {
                toast.error("Email is empty. Please create it again."); 
            } else if(password.length === 0) {
                toast.error("Password is empty. Please create it again."); 
            } else {
                // dispatch register with userData
                dispatch(register(userData)); 
            }; 
        }; 
    }; 

    if(isLoading) {
        return <Spinner />
    }

    return (
        <div>
            <Navbar menuOpen={menuOpen} setMenuOpen={setMenuOpen} />
            <Menu menuOpen={menuOpen} setMenuOpen={setMenuOpen} />            
            <div className="register">
                <div className="register-top">
                    <div className="register-title">
                        <h1>Register</h1>
                        <p>Create your free account</p>
                    </div>
                    <form className="register-form" onSubmit={handleSubmit}>
                        <label
                            htmlFor="username"
                        >
                            Username</label>
                        <input 
                            id="username"
                            name="username"
                            value={username}
                            onChange={handleChange}
                            type="username" 
                            placeholder="Your username" 
                        />
                        <label
                            htmlFor="email"
                        >
                            Email</label>
                        <input 
                            id="email"
                            name="email"
                            value={email}
                            onChange={handleChange}
                            type="email" 
                            placeholder="Your email" 
                        />
                        <label
                            htmlFor="password"
                        >
                            Password</label>
                        <input 
                            id="password"
                            name="password"
                            value={password}
                            onChange={handleChange}
                            type="password" 
                            placeholder="Your password" 
                        />
                        <button>Register</button>
                        <Link to="/login">Already have an account?</Link>
                    </form>
                </div>
            </div>
        </div>
    )
    
}

export default Register