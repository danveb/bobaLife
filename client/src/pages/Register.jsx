import { useState, useEffect } from "react"; 
import { Link, useNavigate } from "react-router-dom"; 
import { useSelector, useDispatch } from "react-redux"; 
import { toast } from "react-toastify"; 
import { register, reset } from "../redux/auth/authSlice"; 
import { Navbar, Menu, Spinner } from "../components/index";
import "./Register.scss"; 

const Register = ({ menuOpen, setMenuOpen }) => {
    // useState 
    const [formData, setFormData] = useState({
        username: "", 
        email: "", 
        password: "",
    }); 

    // destructured form data
    const { username, email, password } = formData; 

    // useNavigate
    const navigate = useNavigate();  

    // useDispatch
    const dispatch = useDispatch(); 

    // select from state.auth
    const { user, isLoading, isError, isSuccess, message } = useSelector((state) => state.auth); 

    // useEffect 
    useEffect(() => {
        if(isError) {
            toast.error(message); 
        }
        // if success or user logged in? 
        if(isSuccess || user) {
            navigate("/")
        }
        // dispatch reset() 
        dispatch(reset()); 

        // dependency array to fire useEffect if there are any changes 
    }, [user, isError, isSuccess, message, navigate, dispatch]);

    // handleChange
    const handleChange = (e) => {
        setFormData((prevState) => ({
            ...prevState, 
            [e.target.name]: e.target.value
        })); 
    }; 

    // handleSubmit
    const handleSubmit = (e) => {
        e.preventDefault(); 
        if(!password) {
            toast.error("Please enter password")
        } else {
            const userData = {
                username, 
                email, 
                password,
            }
            // dispatch register with userData
            dispatch(register(userData)); 
        }; 
    }; 

    if(isLoading) {
        return <Spinner />
    }

    return (
        <div>
            <Navbar menuOpen={menuOpen} setMenuOpen={setMenuOpen} />
            <Menu menuOpen={menuOpen} setMenuOpen={setMenuOpen} />            <div className="register">
                <div className="register-wrapper">
                    <h1>Register</h1>
                    <p>Please enter your details</p>
                    <form onSubmit={handleSubmit} >
                        <label htmlFor="username" />
                        <input
                            id="username" 
                            type="username"
                            name="username"
                            placeholder="Enter a username"
                            value={username} 
                            onChange={handleChange}
                        />
                        <label htmlFor="email" />
                        <input
                            id="email" 
                            type="email"
                            name="email"
                            placeholder="Enter an email"
                            value={email} 
                            onChange={handleChange}
                        />
                        <label htmlFor="password" />
                        <input 
                            id="password"
                            type="password"
                            name="password"
                            placeholder="Enter a password"
                            value={password} 
                            onChange={handleChange}
                        />
                        <button type="submit">Register</button>
                        <Link to="/login">Sign In Here</Link>
                    </form>
                </div>
            </div>
        </div>
    )
    
}

export default Register