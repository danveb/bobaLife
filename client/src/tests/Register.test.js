import { render, screen } from "@testing-library/react"; 
import { store } from "../app/store"; 
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import Register from "../components/Register/Register"; 

// smoke test
test("renders without crashing", () => {
    render(
        <Provider store={store}>
            <BrowserRouter>
                <Register />
            </BrowserRouter>
        </Provider>
    ); 
}); 

// snapshot test
test("matches snapshot", () => {
    const { asFragment } = render(
        <Provider store={store}>
            <BrowserRouter>
                <Register />
            </BrowserRouter>
        </Provider>
    ); 
    expect(asFragment()).toMatchSnapshot(); 
});

// getByText
test("should display header display", () => {
    const { getByText } = render(
        <Provider store={store}>
            <BrowserRouter>
                <Register />
            </BrowserRouter>
        </Provider>
    ); 
    expect(getByText("Create your free account", { exact: true })).toBeInTheDocument(); 
});

// getByLabelText
test("should find username label", () => {
    render(
        <Provider store={store}>
            <BrowserRouter>
                <Register />
            </BrowserRouter>
        </Provider>
    );
    screen.getByLabelText("Username"); 
}); 