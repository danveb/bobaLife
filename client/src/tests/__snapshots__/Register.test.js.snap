// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-caXVBt gJpVKA register"
  >
    <div
      class="sc-hjriPb dWFijc"
    >
      <div
        class="sc-gUAEMC eoOTrv"
      >
        <h1
          class="sc-fWjsSh kVywhz"
        >
          Register
        </h1>
        <p
          class="sc-fHsOPI lcsjjZ"
        >
          Create your free account
        </p>
      </div>
      <form
        class="sc-gzzPqb gPOFHw register-form"
      >
        <label
          class="sc-gkJlnC juEHOj"
          for="username"
        >
          Username
        </label>
        <input
          class="sc-fmrZth gfnrlj"
          id="username"
          name="username"
          placeholder="Your username"
          type="username"
          value=""
        />
        <label
          class="sc-gkJlnC juEHOj"
          for="email"
        >
          Email
        </label>
        <input
          class="sc-fmrZth gfnrlj"
          id="email"
          name="email"
          placeholder="Your email"
          type="email"
          value=""
        />
        <label
          class="sc-gkJlnC juEHOj"
          for="password"
        >
          Password
        </label>
        <input
          class="sc-fmrZth gfnrlj"
          id="password"
          name="password"
          placeholder="Your password"
          type="password"
          value=""
        />
        <button
          class="sc-eXBvqI bsbuvm"
        >
          Register
        </button>
        <a
          href="/login"
        >
          Already have an account?
        </a>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
